packages/playwright-test/src/matchers/expect.ts

```diff
@@ -267,17 +272,28 @@ class ExpectMetaInfoProxyHandler implements ProxyHandler<any> {
         step.complete({});
       };
 
-      try {
-        const expectZone: ExpectZone = { title: defaultTitle, wallTime };
-        const result = zones.run<ExpectZone, any>('expectZone', expectZone, () => {
-          return matcher.call(target, ...args);
-        });
-        if (result instanceof Promise)
-          return result.then(() => finalizer()).catch(reportStepError);
-        else
+      const expectZone: ExpectZone = { title: defaultTitle, wallTime };
+      // Process the async matchers separately to preserve the zones in the stacks.
+      if (this._info.isPoll || matcherName in customAsyncMatchers) {
+        return (async () => {
+          try {
+            await zones.run<ExpectZone, any>('expectZone', expectZone, async () => {
+              await matcher.call(target, ...args);
+            });
+            finalizer();
+          } catch (e) {
+            reportStepError(e);
+          }
+        })();
+      } else {
+        try {
+          zones.run<ExpectZone, any>('expectZone', expectZone, () => {
```

Comment:
I guess we don't really need the zone here, since we don't run our own matchers?

Reply:
You are right!